packages/cozy-client/src/associations/Association.js

```diff
@@ -225,7 +225,7 @@ class Association {
    *
    * @returns {QueryDefinition | QueryState}
    */
-  static query() {
+  static query(document, client, assoc) {
```

Comment:
Those parameters can be typed:

client is a CozyClient
document is a CozyClientDocument
assoc is an Association I guess

Reply:
Ok I just made this.
Note that I had to propagate those definitions to those so TS is able to detect the override :

- HasMany.js
- HasManyFiles.js
- HasManyInPlace.js
- HasOne.js
- HadOneInPlace.js

Also TS detected that the return type was not `{QueryDefinition | QueryState}` but `{CozyClientDocument | QueryDefinition}`. I made the change but I didn't check if all callers expect that type (not sure how to easily find them all as `query` is kind of generic term)

Reply:
Ok, now there is [a problem with tests](https://travis-ci.com/github/cozy/cozy-client/builds/229340890).

Importing `CozyClient` in those files breaks a lot of tests (`watch` and `build` seem to work correctly though). I'm not sure why, maybe recursive import is not handled correctly?

That being said, i don't know how to type `client` as `CozyClient` without breaking everything. Any idea @ptbrowne @Crash-- ?

Reply:
Seems like a circular dependency bug, cozy-client imports association that imports hasmany that imports cozy-client. HasMany here seems to be imported first, which means cozy-client is not ready yet when the subclassing done. 

You can try using `import` inside the type annotation so that the import does not happen in JS context but only in typescript context.

```
  /**
   * @param {typeof import('../CozyClient').default} client - A client
   */
```

See https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#import-types

Reply:
Sadly this syntax is not supported by our jsdoc version (see https://github.com/microsoft/TypeScript/issues/14377). `yarn test` will work, but not `yarn docs`.
There is a fix for this but it doesn't seems to be merged in jsdoc yet : https://github.com/ricardohbin/jsdoc/pull/1 

I temporarily typed `CozyClient` as `object` so I can merge this PR and I will open an issue.

Reply:
Ah yes, this is unfortunate, good call for the temporary typing. I think jsdoc should be removed and tsdoc should be used. This way there would no incompatibility between different parsers. I have not used it though, and do not have any XP on it.